src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,604 @@
+
+/// Serialization of `GetNodes`. Resulting length should be
+/// [`GET_NODES_SIZE`](./constant.GET_NODES_SIZE.html).
+impl ToBytes for GetNodes {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_slice!(self.pk.as_ref()) >>
+            gen_be_u64!(self.id)
+        )
+    }
+}
+
+/** De-serialization of bytes into `GetNodes`. If less than
+[`GET_NODES_SIZE`](./constant.GET_NODES_SIZE.html) bytes are provided,
+de-serialization will fail, returning `None`.
+*/
+impl FromBytes for GetNodes {
+    named!(from_bytes<GetNodes>, do_parse!(
+        pk: call!(PublicKey::parse_bytes) >>
+        id: ne_u64 >>
+        (GetNodes { pk: pk, id: id })
+    ));
+}
+
+/** Response to [`GetNodes`](./struct.GetNodes.html) request, containing up to
+`4` nodes closest to the requested node.
+
+Packet type `0x04`.
+
+Serialized form:
+
+Length      | Contents
+----------- | --------
+`1`         | Number of packed nodes (maximum 4)
+`[39, 204]` | Nodes in packed format
+`8`         | Ping ID
+
+An IPv4 node is 39 bytes, an IPv6 node is 51 bytes, so the maximum size is
+`51 * 4 = 204` bytes.
+
+Serialized form should be put in the encrypted part of DHT packet.
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SendNodes {
+    /** Nodes sent in response to [`GetNodes`](./struct.GetNodes.html) request.
+
+    There can be only 1 to 4 nodes in `SendNodes`.
+    */
+    pub nodes: Vec<PackedNode>,
+    /// Ping id that was received in [`GetNodes`](./struct.GetNodes.html)
+    /// request.
+    pub id: u64,
+}
+
+/// Method assumes that supplied `SendNodes` has correct number of nodes
+/// included – `[1, 4]`.
+impl ToBytes for SendNodes {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_slice!(self.nodes.as_ref()) >>
+            gen_be_u64!(self.id)
+        )
+    }
+}
+
+/** Method to parse received bytes as `SendNodes`.
+
+    Returns `None` if bytes can't be parsed into `SendNodes`.
+*/
+impl FromBytes for SendNodes {
+    named!(from_bytes<SendNodes>, do_parse!(
+        nodes_number: le_u8 >>
+        nodes: cond_reduce!(
+            nodes_number > 0 && nodes_number <= 4,
+            count!(PackedNode::parse_bytes, nodes_number as usize)
+        ) >>
+        id: ne_u64 >>
+        (SendNodes {
+            nodes: nodes,
+            id: id
+        })
+    ));
+}
+
+/** Standard DHT packet that encapsulates in the encrypted payload
+[`DhtPacketT`](./trait.DhtPacketT.html).
+
+Length      | Contents
+----------- | --------
+`1`         | `uint8_t` [`PacketKind`](../packet_kind/enum.PacketKind.html)
+`32`        | Sender DHT Public Key
+`24`        | Random nonce
+variable    | Encrypted payload
+
+`PacketKind` values for `DhtPacket` can be only `<= 4`.
+
+https://zetok.github.io/tox-spec/#dht-packet
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DhtPacket {
```

Comment:
Also you have to parse packets with tags inside the packet. Hence https://github.com/zetok/tox/blob/master/src/toxcore/tcp/packet.rs#L33..L92  .